Lambda to OpenFaaS Blog Post

burtonr opened the pull request:
Blog post to describe taking an existing AWS Lambda function and converting it to an OpenFaaS function

For reference, the background image was obtained free for use with no attribution required from [Pixabay](https://pixabay.com/photos/glacier-national-park-geese-over-lake-4331626/)

burtonr commented:
@alexellis This is ready for a review.

I'll need to update the post date once it's approved and we decide on when it should be published. It's currently set in the past so it can be visible for local testing.

LucasRoesler commented on _posts/2019-09-20-lambda-to-openfaas.md:
```diff
@@ -0,0 +1,282 @@
+#### Use secrets
+We're not moving all of our infrastructure off of AWS, so we'll continue to use DynamoDB. Again, since we want this function to be portable, we'll need to be able to use our AWS Access Key ID and AWS Secret Access Key associated with the same IAM account that our Lambda function was using.
+
+First, you'll need to create a user in AWS with the role that the Lambda function was using. Check the box for "Programmatic access" so that the access keys will be generated. 
```
Are there any specific docs or AWS walkthroughs we can also link to. The aws UI can be intimidating for some people. Being able to link to another guide could be useful. Especially, if there is one specific for dynamodb

LucasRoesler commented on _posts/2019-09-20-lambda-to-openfaas.md:
```diff
@@ -0,0 +1,282 @@
+- Replaced the AWS specific response object with the template's `context`
+- Set the `context.success` and `context.fail` status codes and values
+
+If you look at the `git diff --stat` output, we actually removed more code than we added!
```
With all these changes, is the function still deployable on lambda? Might be worth mentioning

burtonr replied:
Added a note mentioning that because of the changes made, it is no longer able to run on Lambda

LucasRoesler commented on _posts/2019-09-20-lambda-to-openfaas.md:
````diff
@@ -0,0 +1,282 @@
+$ mv shortener.yml stack.yml
+```
+
+Now, copy the `redirector` properties from the `redirecter.yml` file, and paste them in below the `shortener` function property.
````
What about the append flag?

burtonr replied:
Good idea! 

I moved this piece up to the "create a function" section adding the `--append` flag to the steps.

LucasRoesler commented on _posts/2019-09-20-lambda-to-openfaas.md:
```diff
@@ -0,0 +1,282 @@
+## Wrapping Up
+We've migrated our function code from being restricted to only deployable on AWS Lambda infrastructure to being completely portable by using OpenFaaS. 
+
+Additionally, this means that developers are able to test thier functions by deploying to a local Kubernetes cluster using Docker for Windows, Docker for Mac, [microk8s](https://johnmccabe.net/technology/projects/openfaas-on-microk8s/), or [KinD](https://blog.alexellis.io/be-kind-to-yourself/)
```
Typo in thier

burtonr replied:
good eye!

rgee0 commented on _posts/2019-09-20-lambda-to-openfaas.md:
````diff
@@ -0,0 +1,282 @@
+```
+Now, create the new "shortener" function:
+```shell
+$ faas-cli new shortener --lang node10-express
````
should we use `--prefix` here, or set the env var?  It'll save mentioning the image prefix in the note later.

alexellis replied:
+1

burtonr replied:
Good idea! I added a bit about using the `DOCKER_USER` env var that makes the rest of the commands shorter

rgee0 commented on _posts/2019-09-20-lambda-to-openfaas.md:
```diff
@@ -0,0 +1,282 @@
+    secrets:
+      - shorturl-dynamo-key
+      - shorturl-dynamo-secret
+  redirecter:
```
We've got an inconsistency from 205 downwards about how we are spelling `redirector`

alexellis replied:
+1 I think `redirector` looks better

burtonr replied:
Good eye!

rgee0 commented on _posts/2019-09-20-lambda-to-openfaas.md:
````diff
@@ -0,0 +1,282 @@
+$ echo $PASSWORD | faas-cli login -u admin -s
+```
+
+- Now, we're ready to build the container images.
````
Since weve used `stack.yml` I'd consider using `up` as the focus, rather than a footnote.

alexellis replied:
`-u admin` is not required. instead of `-s` use the full flag name.

burtonr replied:
Updated to use the `up` command, and mentioned that it includes the other 3 commands as 1.

Also updated the login command to use the full flag name

alexellis commented on _posts/2019-09-20-lambda-to-openfaas.md:
````diff
@@ -0,0 +1,282 @@
+
+Let's have a look at the `shortener` function which creates and stores the shortened URLs:
+
+```javascript
````
I didn't look yet, but is this in a GitHub repo ?

burtonr replied:
Yes, there is a link in the prerequisites mentioning "Existing AWS Lambda functions". 

https://github.com/burtonr/lambda-openfaas-blog

alexellis replied:
Can you say what the filename is under each code snippet and make it a hyperlink?

burtonr commented:
Thank you all for the great feedback! 
As we discussed on Slack, I will be putting this PR on hold for now to create a slightly different version of this post that focuses on the simplest function from the Lambda docs for the "hello world" function. This will then get updated as "Part 2", getting in to a little more functionality/changes. Eventually, a "Part 3" that could show moving entirely away from AWS showcasing the portability of OpenFaaS.

alexellis commented:
The feedback on Slack to write a simple example in the introduction showing the difference between the OpenFaaS and Lambda signature.

Given that you have already done all of the work necessary to port a function I don't see why this would put the post on hold?

What was it that you understood?

burtonr commented:
Apologies @alexellis. I understood it as a "3 part blog", as in 3 different posts, rather than 3 different _sections_ of the same post.

It's not a problem, I can make the updates from the feedback here, and let you all know once the updates are pushed.

alexellis commented:
That's two sections, what's the third?

burtonr commented:
Post 3 would have been migrating entirely out of AWS to a different provider. Something to showcase the portability of the OpenFaaS platform once all the functions were moved.

burtonr commented on _posts/2019-09-20-lambda-to-openfaas.md:
```diff
@@ -0,0 +1,282 @@
+dark_background: false
+---
+
+In this post, I'll walk you through the steps taken to migrate existing AWS Lambda functions to OpenFaaS compatible functions in order to allow easier local testing, familiar development process, and the freedom to use any number of other cloud providers without any code changes.
```
Good catch! Moved.

burtonr commented on _posts/2019-09-20-lambda-to-openfaas.md:
```diff
@@ -0,0 +1,282 @@
+
+and paste it in to the `handler.js` file, replacing everything below the line: `module.exports = (event, context) => {`
+
+> You'll also need to copy over the `require` lines, and any `const` you've defined above the `exports.handler` line.
```
Added link to the file on Github

burtonr commented on _posts/2019-09-20-lambda-to-openfaas.md:
````diff
@@ -0,0 +1,282 @@
+```
+
+#### Use secrets
+We're not moving all of our infrastructure off of AWS, so we'll continue to use DynamoDB. Again, since we want this function to be portable, we'll need to be able to use our AWS Access Key ID and AWS Secret Access Key associated with the same IAM account that our Lambda function was using.
````
Added a sentence in the introduction explaining that only the functions will be moved, and made reference to it in the "secrets" section

burtonr commented:
@LucasRoesler, @rgee0, and @alexellis I've updated this post with all of your suggestions.

I also added a "Getting Started" section that shows the small differences between the 2 "hello world" functions.

Please let me know what you think. Ignore my previous comment about putting this on hold, I had misunderstood our Slack conversation

alexellis commented on _posts/2019-09-20-lambda-to-openfaas.md:
```diff
@@ -0,0 +1,314 @@
+
+In this post, We'll walk through the steps taken to migrate existing AWS Lambda functions to OpenFaaS compatible functions in order to allow easier local testing, familiar development process, and the freedom to use any number of other cloud providers without any code changes. We won't be moving all of the infrastructure, only the functions. We'll keep the DynamoDB table as well as the IAM roles/ ...
+
+Without further ado, let's see what it takes to upgrade your AWS Lambda functions to OpenFaaS and Serverless 2.0!
```
`to Serverless 2.0 using OpenFaaS.`

alexellis commented on _posts/2019-09-20-lambda-to-openfaas.md:
```diff
@@ -0,0 +1,314 @@
+
+[TLS & Custom Domains for your Functions](/blog/custom-domains-function-ingress/)
+
+mTLS, Traffic Splitting, and more with [OpenFaaS and Linkerd2](https://github.com/openfaas-incubator/openfaas-linkerd2)
```
What's the link between Lambda and mTLS etc?

burtonr replied:
There's no direct link. I added it as a link to show what other features are available now that the functions are on OpenFaaS. Something that's not possible with Lambda (as far as I know)

alexellis commented on _posts/2019-09-20-lambda-to-openfaas.md:
```diff
@@ -0,0 +1,314 @@
+
+[Build your own OpenFaaS Cloud with AWS EKS](/blog/eks-openfaas-cloud-build-guide)
+
+[TLS & Custom Domains for your Functions](/blog/custom-domains-function-ingress/)
```
Let's use the faas-lambda post instead for this bullet

burtonr replied:
I would prefer to keep this, as custom domains for Lambda is possible with the API Gateway, as is HTTPS, and may be an important feature for some. In the post, the URL is shown as `http://localhost:8080/function/<function name>` and that may not be appealing. This link just shows that OpenFaaS also supports those features/

Happy to replace the Linkerd link with the faas-lambda though since it's more closely related

alexellis commented on _posts/2019-09-20-lambda-to-openfaas.md:
```diff
@@ -0,0 +1,314 @@
+> The `up` command combines the `build`, `push`, and `deploy` commands into one cohesive operation.
+
+
+#### Invoking the Function
```
I don't get what's happening here? I feel like you need to explain it more?

burtonr replied:
Didn't realize how empty that really was. Thanks for pointing that out! 

I've added a bit more explanation as well as a mention of the async functions. Let me know if you think there should be more, or less.

alexellis commented on _posts/2019-09-20-lambda-to-openfaas.md:
```diff
@@ -0,0 +1,330 @@
+dark_background: false
+---
+
+AWS introduced the world to the idea of serverless architecture back in 2014 with the announcement of Lambda. The idea was that you could upload a zip file containing your code logic, and AWS would manage the infrastructure behind the scenes, billing you only for the time the code was being executed.
```
> AWS introduced the world to the idea of serverless architecture back in 2014 with the announcement of Lambda.

Did they? I understood that Serverless was being talked about before then, how about referring to the product launch and linking to it.

burtonr replied:
Good point. They launched the feature. Updated the wording and linked to the re:Invent talk

alexellis commented on _posts/2019-09-20-lambda-to-openfaas.md:
```diff
@@ -0,0 +1,330 @@
+dark_background: false
+---
+
+AWS introduced the world to the idea of serverless architecture back in 2014 with the announcement of Lambda. The idea was that you could upload a zip file containing your code logic, and AWS would manage the infrastructure behind the scenes, billing you only for the time the code was being executed.
```
"The idea was that you could upload a zip file containing your code logic, and AWS would manage the infrastructure behind the scenes, billing you only for the time the code was being executed."

Break the sentence up using multiple sentences or and. I.e. stop at "behind the scenes" and start a new sentence

burtonr replied:
Done

alexellis commented on _posts/2019-09-20-lambda-to-openfaas.md:
```diff
@@ -0,0 +1,330 @@
+
+AWS introduced the world to the idea of serverless architecture back in 2014 with the announcement of Lambda. The idea was that you could upload a zip file containing your code logic, and AWS would manage the infrastructure behind the scenes, billing you only for the time the code was being executed.
+
+In his talk [Welcome to Serverless 2.0](https://www.youtube.com/watch?v=JvXm-oHi5Mg), Alex Ellis described the new standard for serverless functions: Docker/OCI images exposing a port for execution, deployed to a Kubernetes cluster. This makes your functions completely portable and removes many of the limitations of the previous generation of serverless functions.
```
"the new standard" -> "a new standard"

alexellis commented on _posts/2019-09-20-lambda-to-openfaas.md:
```diff
@@ -0,0 +1,330 @@
+
+AWS introduced the world to the idea of serverless architecture back in 2014 with the announcement of Lambda. The idea was that you could upload a zip file containing your code logic, and AWS would manage the infrastructure behind the scenes, billing you only for the time the code was being executed.
+
+In his talk [Welcome to Serverless 2.0](https://www.youtube.com/watch?v=JvXm-oHi5Mg), Alex Ellis described the new standard for serverless functions: Docker/OCI images exposing a port for execution, deployed to a Kubernetes cluster. This makes your functions completely portable and removes many of the limitations of the previous generation of serverless functions.
```
"Docker/OCI images exposing a port for execution, deployed to a Kubernetes cluster" => the "Serverless 2.0 runtime contract" (link to the docs page)

burtonr replied:
You're referring to the "Workloads" document right? I don't think we have anything explicitly calling out the "Serverless 2.0 contract"

alexellis commented on _posts/2019-09-20-lambda-to-openfaas.md:
```diff
@@ -0,0 +1,330 @@
+
+In his talk [Welcome to Serverless 2.0](https://www.youtube.com/watch?v=JvXm-oHi5Mg), Alex Ellis described the new standard for serverless functions: Docker/OCI images exposing a port for execution, deployed to a Kubernetes cluster. This makes your functions completely portable and removes many of the limitations of the previous generation of serverless functions.
+
+In this post, We'll walk through the steps taken to migrate existing AWS Lambda functions to OpenFaaS compatible functions in order to allow easier local testing, familiar development process, and the freedom to use any number of other cloud providers without any code changes. We won't be moving all of the infrastructure, only the functions. We'll keep the DynamoDB table as well as the IAM roles/ ...
```
First "We'll" should be "we'll"

alexellis replied:
Say it positive, not negative:

"We won't be moving all of the infrastructure, only the functions."

=> 

"I'll show you how to migrate the compute to OpenFaaS, whilst continuing to use AWS services such as DynamoDB."

alexellis commented on _posts/2019-09-20-lambda-to-openfaas.md:
```diff
@@ -0,0 +1,330 @@
+
+In this post, We'll walk through the steps taken to migrate existing AWS Lambda functions to OpenFaaS compatible functions in order to allow easier local testing, familiar development process, and the freedom to use any number of other cloud providers without any code changes. We won't be moving all of the infrastructure, only the functions. We'll keep the DynamoDB table as well as the IAM roles/ ...
+
+Without further ado, let's see what it takes to upgrade your AWS Lambda functions to OpenFaaS and Serverless 2.0 using OpenFaaS!
```
Delete "Without further ado, "

Delete "!"

alexellis commented on _posts/2019-09-20-lambda-to-openfaas.md:
```diff
@@ -0,0 +1,330 @@
+
+## Prerequisites
+
+- Existing AWS Lambda functions (optional)
```
An existing

alexellis commented on _posts/2019-09-20-lambda-to-openfaas.md:
```diff
@@ -0,0 +1,330 @@
+- OpenFaaS CLI
+  - Check the documentation for [installation instructions](https://docs.openfaas.com/cli/install/)
+- IDE, or other text editor
+  - I recommend [Visual Studio Code](https://code.visualstudio.com/)
```
Shorter -> "A text editor, I recommend VSCode"

alexellis commented on _posts/2019-09-20-lambda-to-openfaas.md:
````diff
@@ -0,0 +1,330 @@
+};
+```
+
+Similarly, when you create a new function with the OpenFaaS CLI and the `node10-express` template, sample code is provided as shown here:
````
Replace "similarly" for something else. It can be hard for International speakers to say.

"In a similar way" or "Let's examine the example for OpenFaaS"
...

"See how they are so similar?"

burtonr replied:
Thank you for pointing that out!

alexellis commented on _posts/2019-09-20-lambda-to-openfaas.md:
```diff
@@ -0,0 +1,330 @@
+
+module.exports = (event, context) => {
+    let err;
+    const result =             {
```
` =             {` why the whitespace?

burtonr replied:
Bad copy... good catch

alexellis commented on _posts/2019-09-20-lambda-to-openfaas.md:
````diff
@@ -0,0 +1,330 @@
+}
+```
+
+At a quick glance, the two function samples are very similar. The main difference is in the invocation parameters. OpenFaaS passes in the `context` object which is used to provide additional context to the result. The context includes the properties `status`, `succeed`, `error`, and others that make discovering functionality and managing the response of the function easy without having to dig thr ...
````
"At a quick glance," -> "At a glance,"

alexellis commented on _posts/2019-09-20-lambda-to-openfaas.md:
```diff
@@ -0,0 +1,330 @@
+At a quick glance, the two function samples are very similar. The main difference is in the invocation parameters. OpenFaaS passes in the `context` object which is used to provide additional context to the result. The context includes the properties `status`, `succeed`, `error`, and others that make discovering functionality and managing the response of the function easy without having to dig thr ...
+
+As you can see, migrating the sample function from Lambda to OpenFaaS would be as easy as adding the `context` object to the invocation, and moving the values from the inline `response` object to the OpenFaaS included `context` object properties.
+
```
`You can see the documentation and other examples for the node10-express template here.` <- add this and a link

alexellis commented on _posts/2019-09-20-lambda-to-openfaas.md:
```diff
@@ -0,0 +1,330 @@
+
+## AWS Lambda
+
+The Lambda functions we'll be migrating are part of a URL shortener. One function creates and stores the shortened URL in DynamoDB, and the other function is responsible for looking up the destination URL, and redirecting the browser.
```
Explain to the reader why are there two functions and not one. What benefit is there? Does it make sense?

burtonr replied:
Explained that the reason would be to put the creation function in a private network to limit access. In all of the tutorials/blog posts I found, this is the recommended approach.

alexellis replied:
Please can you explain this to the reader in the blog post?

alexellis replied:
Edit 2 ^

alexellis commented on _posts/2019-09-20-lambda-to-openfaas.md:
```diff
@@ -0,0 +1,330 @@
+    });
+}
+
+function createAndStorePath(originalUrl, callback) {
```
How are you authenticating clients that have write access to the API?

burtonr replied:
I skipped the authentication mostly to simplify the post. I did add a sentence mentioning that the "shortener" creation function could be accessible in a private network to limit the access

alexellis replied:
Please can you mention this in the blog post and suggest how you may use authentication? Leaving it open is OK for development, may be unwise to promote on the blog without giving context.

alexellis replied:
Edit 1 ^

burtonr commented:
I've just added a section about using the EKS IAM Service Account. Thanks to @ewilde for pointing that out.
